File: py_drill/Problems/TR10_AllNodesDistanceKInBinaryTree.py
# REF  : https://leetcode.com/problems/all-nodes-distance-k-in-binary-tree/
# NOTES :
#       1st way : find node address if not given
#                 create node:parent hash
#                 apply BFS (iterative using queue) with a visited set
#       2nd way : Convert this tree to graph and DFS with visit set till depth k.

from collections import deque

class TreeNode:
    def __init__(self, x):
        self.val = x
        self.left = None
        self.right = None

# T: O(2n)
# S: O(2n)
class Solution:
    def distanceK(self, root: TreeNode, target: TreeNode, k: int) -> list[int]:

        parentHash = {}
        def createParentHash(root):
            if root is None:
                return

            if root.left:
                parentHash[root.left] = root
            if root.right:
                parentHash[root.right] = root

            createParentHash(root.left)
            createParentHash(root.right)

        if root is None:
            return []

        createParentHash(root)

        result = []
        visited = set()
        q = deque()

        q.append(target)
        while q:
            for i in range(len(q)):
                node = q.popleft()
                visited.add(node)
                if k == 0:                          # This 'if' will be removed if 'print nodes till k' is asked
                    result.append(node.val)

                if node in parentHash and parentHash[node] not in visited:
                    q.append(parentHash[node])

                if node.left and node.left not in visited:
                    q.append(node.left)

                if node.right and node.right not in visited:
                    q.append(node.right)

            k -= 1
            if k < 0:
                break

        return result


# Need to fix it
class Solution2:
    def distanceK(self, root: TreeNode, target: TreeNode, k: int) -> list[int]:

        result = []

        def printNodeAtk(root, target, k):
            def printSubtree(root, k):
                if root is None or k < 0:
                    return
                if k == 0:
                    result.append(root.val)

                printSubtree(root.left, k - 1)
                printSubtree(root.right, k - 1)

            if root is None:
                return -1

            if root == target:
                printSubtree(target, k)
                return result

            dl = printNodeAtk(root.left, target, k)
            if dl != -1:
                if dl + 1 == k:
                    result.append(root.val)
                else:
                    printSubtree(root.right, k-dl-2)

                return dl + 1

            dr = printNodeAtk(root.right, target, k)
            if dr != -1:
                if dr + 1 == k:
                    result.append(root.val)
                else:
                    printSubtree(root.left, k - dr - 2)

                return dr + 1


        ret = printNodeAtk(root, target, k)
        if ret == -1:
            return result


def main():
    pass

if __name__ == "__main__":
    main()

File: cpp_drill/Problems/TR10_AllNodesDistanceKInBinaryTree.cpp
#include <iostream>
#include <vector>
#include <unordered_map>
#include <queue>
using namespace std;

// REF : https://leetcode.com/problems/all-nodes-distance-k-in-binary-tree/
// NOTES : 
//      Steps for this solution -
//      1st Solution -
//      	Find node address if not given (Optional)
//      	Create node : parent Hashmap (DFS)
//      	Apply BFS with visited set        (BFS)
//      2nd Solution -
//      	Convert this tree to graph
//      	DFS with visit set till depth k

/**
 * Definition for a binary tree node.
 * struct TreeNode {
 *     int val;
 *     TreeNode *left;
 *     TreeNode *right;
 *     TreeNode(int x) : val(x), left(NULL), right(NULL) {}
 * };
 */

// T: O(n)
// S: O(n)
class Solution {
public:

    // For creating node : parent Hashmap
    void dfs(TreeNode* node, unordered_map<TreeNode*, TreeNode*>& mp) {
        if (node == nullptr) return;

        if (node->left) mp[node->left] = node;
        if (node->right) mp[node->right] = node;

        dfs(node->left, mp);
        dfs(node->right, mp);
    }

    vector<int> distanceK(TreeNode* root, TreeNode* target, int k) {
        unordered_map<TreeNode*, TreeNode*> parentHash;
        dfs(root, parentHash);

        vector<int> result;
        queue<TreeNode*> q;
        unordered_map<TreeNode*, bool> visited;

        // Need to start BFS from the target node
        // NOTE : Usually while(len--)/for loop is not required in bfs
        //        Only put here because we want to identify each level
        
        q.push(target);
        while (!q.empty()) {
            int len = q.size();
            while (len--) {                             // for loop also fine
                TreeNode* node = q.front(); q.pop();
                visited[node] = true;

                // At k==0 level, all nodes will be added to the result
                if (k==0) result.push_back(node->val);

                if (parentHash.find(node) != parentHash.end() &&
                    visited[parentHash[node]] == false)
                    q.push(parentHash[node]);

                if (node->left && visited[node->left] == false)
                    q.push(node->left);

                if (node->right && visited[node->right] == false)
                    q.push(node->right);
            }
            k--;
            if (k < 0) break;
        }

        return result;
    }
};

// T: O(n)
// S: O(n)
class Solution {
public:

    // Converting Tree into Graph
    // directed acyclic graph to undirected acyclic graph
    void dfs(TreeNode* node, unordered_map<TreeNode*, vector<TreeNode*>>& g) {
        if (node == nullptr) return;

        if (node->left) {
            g[node].push_back(node->left);
            g[node->left].push_back(node);
        }
        if (node->right) {
            g[node].push_back(node->right);
            g[node->right].push_back(node);
        }

        dfs(node->left, g);
        dfs(node->right, g);
    }

    vector<int> distanceK(TreeNode* root, TreeNode* target, int k) {
        unordered_map<TreeNode*, vector<TreeNode*>> graph;
        dfs(root, graph);

        vector<int> result;
        queue<TreeNode*> q;
        unordered_map<TreeNode*, bool> visited;

        // Need to start BFS from the target node
        q.push(target);
        while (!q.empty()) {
            int len = q.size();
            while (len--) {
                TreeNode* node = q.front(); q.pop();
                visited[node] = true;

                // At k==0 level, all nodes will be added to the result
                if (k==0) result.push_back(node->val);

                for (TreeNode* nei : graph[node])
                    if (visited[nei] == false)
                        q.push(nei);
            }
            k--;
            if (k < 0) break;
        }

        return result;
    }
};